docs/Blog-Post.md

```diff
@@ -0,0 +1,276 @@
+* **Efficiency**: We wanted to achieve the other three goals without sacrificing (much) efficiency.
+
+We chose [Jax](https://github.com/google/jax/) as our framework because it is a powerful, flexible, and performant,
+and offers strong reproducibility guarantees. Jax works well on TPUs, while we found that PyTorch support was still uneven.
+Jax is also a natural choice because it allows you to focus on the "what" of your code, and not on the "how:" details of
+partitioning and communication can be left to the XLA compiler. Finally, Jax makes reproducibility easy, since it uses
+bitwise deterministic PRNGs by default, with careful control over the PRNG state.
+
+However, Jax is a low-level framework, and we found that, by itself, it did not provide the legibility that we wanted.
+
+For us, legibility is a top concern. We therefore created two new libraries: **Haliax** and **Levanter**. Haliax is a
+named tensor library that focuses on improving the legibility and compositionality of deep learning code while still
+maintaining efficiency and scalability. Levanter is a library for training foundation models built on top of Haliax. In
+addition to the goals of legibility, efficiency, and scalability, Levanter further strives for bitwise reproducibility,
+meaning that the same code with the same data will produce the exact same result, even in the presence of preemption and
+restarts.
+
+## Haliax: Legibility via Named Tensors
+
+Haliax is a Jax library for building neural networks with named tensors, built on Jax and [Equinox](https://github.com/patrick-kidger/equinox),
+which is a neural network library built on Jax that provides a familiar, PyTorch-like module structure.
+
+Named Tensors are a powerful abstraction that allow you to give names to the axes of your tensors. These names help
+make your code more legible, more composable, and avoid bugs. In Haliax, they also form the basis of how we handle
+scale with [Fully-Sharded Data Parallel](https://engineering.fb.com/2021/07/15/open-source/fsdp/). (See below for our tutorial)
+
+Haliax is modeled on Alexander Rush's [Tensor Considered Harmful](https://nlp.seas.harvard.edu/NamedTensor). In particular, he argues that:
+
+* Named axes are more semantically meaningful and rely less on bitrot-prone comments.
+* Broadcasting leads to unreadable strings of `view`s and `squeeze`s. (To this, I'd add implicit broadcasting is a source of bugs.)
+* Named axes allow you to abstract over unreferenced dimensions, making code more flexible for things like multi-headed attention or flexible attention masking.
+
+Let's take a look at a practical example of how Haliax can be used.
+Here's a minimal attention module implementation in Haliax. For a more detailed introduction,
+please see the [Haliax tutorial](https://colab.research.google.com/drive/1TiTcQQ4V5mopbgCu1SVl-oqJtXn7rFnC).
+
+```python
```

Comment:
I wonder if this is too much code, and there's smaller (1-3 line) snippets that one can put to highlight particular aspects like the named axes...can put '...' for intervening code.

Reply:
think i could just drop the module and just do the axes + attention function?